### DISCUSSION TITLE
posts/Go/interface/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#361] [Announcements]
https://www.liwenzhou.com/posts/Go/12_interface/ 

 李文周的Blog Go语言接口 interface 面向接口编程

**Comment by 272909106** [comment]
package main

import (
	"fmt"
	"os"
)

type logSystem interface {
	consol(interface{})
	wfile(interface{})
}
type localLog struct{}

func (l localLog) consol(log interface{}) {
	loginfo := log.(string) + "\n"
	os.Stdout.WriteString(loginfo)
}
func (l localLog) wfile(log interface{}) {
	logMsg := log.(string) + "\n"
	_, err := os.Stat("interface.log")
	if err != nil {
		fmt.Println("文件不存在!需要创建！")
		// fmt.Println(err)
		file, err := os.OpenFile("interface.log", os.O_APPEND|os.O_CREATE|os.O_EXCL, 0664) //|os.O_CREATE|os.O_EXCL
		if err != nil {
			// fmt.Println(err)
			fmt.Printf("创建打开文件，报错：%s", err)
		}
		defer file.Close()
		if _, err := file.WriteString(logMsg); err == nil {
			fmt.Println("写入文件成功！")
		} else {
			// fmt.Println("文件写入报错！")
			// fmt.Println(err)
			fmt.Printf("文件写入报错！，报错信息：%s", err)
		}
	} else {
		fmt.Println("文件存在！")
		// fmt.Println(fileStat)
		file, err := os.OpenFile("interface.log", os.O_APPEND, 0664) //|os.O_CREATE|os.O_EXCL
		if err != nil {
			fmt.Printf("打开文件错误，报错：%s", err)
		}
		defer file.Close()
		if _, err := file.WriteString(logMsg); err == nil {
			fmt.Println("写入文件成功！")
		} else {
			fmt.Println("文件写入报错！")
			fmt.Println(err)
		}
	}

}

func logger(log string) {
	//  log := log
	var logger logSystem
	var application = localLog{}
	logger = application
	logger.consol(log)
	logger.wfile(log)

}
func main() {
	var word string = "阿里巴巴与四十大盗111!"
	logger(word)
}

**Comment by asd6335683** [comment]
面试题我服了，func (stu *Stduent)中把Student 写成Stduent，var peo People = Student{}加个var peo People = &Student{}就行了

**Comment by cuteculu** [comment]
面试题看了半天，怎么写都报错，明明写的是对的，结果题就是错的，也是够了。

**Comment by Q1mi** [comment]
> @crazypersonal
> 面试题看了半天，怎么写都报错，明明写的是对的，结果题就是错的，也是够了。

我博客上的问题不就是问下面的代码能不能通过编译吗？

**Comment by cuteculu** [comment]
大概是表述有点问题，面试题Speak方法里面的Student 写成了Stduent，所以题有点问题。编译报错是因为要传指针。

**Comment by Q1mi** [comment]
> @crazypersonal
> 大概是表述有点问题，面试题Speak方法里面的Student 写成了Stduent，所以题有点问题。编译报错是因为要传指针。

笔误写错了student这个单词，晚上回去改下。题目的描述确实可以再清晰一下，晚上回去改一下。感谢反馈

**Comment by miniyk2012** [comment]
```
package main

import (
	"fmt"
	"io"
	"os"
)

// 使用接口的方式实现一个既可以往终端写日志也可以往文件写日志的简易日志库。
type Logger interface {
	Info(string)
}

type FileLogger struct {
	filename string
}

func (fl *FileLogger) Info(msg string) {
	var f *os.File
	var err1 error
	if checkFileIsExist(fl.filename) { //如果文件存在
		f, err1 = os.OpenFile(fl.filename, os.O_APPEND|os.O_WRONLY, 0666) //打开文件
		fmt.Println("文件存在")
	} else {
		f, err1 = os.Create(fl.filename) //创建文件
		fmt.Println("文件不存在")
	}
	defer f.Close()
	n, err1 := io.WriteString(f, msg+"\n") //写入文件(字符串)
	if err1 != nil {
		panic(err1)
	}
	fmt.Printf("写入 %d 个字节\n", n)
}

func checkFileIsExist(filename string) bool {
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		return false
	}
	return true
}

type ConsoleLogger struct {
}

func (cl *ConsoleLogger) Info(msg string) {
	fmt.Println(msg)
}

func homework() {
	var logger Logger
	fileLogger := &FileLogger{"log.txt"}
	logger = fileLogger
	logger.Info("Hello")
	logger.Info("how are you")

	consoleLogger := &ConsoleLogger{}
	logger = consoleLogger
	logger.Info("Hello")
	logger.Info("how are you")
}

func main() {
	homework()
}

```

**Comment by shaoxiaUdropUUNshort** [comment]
package main

import (
	"fmt"
	"os"
)

type writer interface {
	info(msg string)
}

type writeToControl struct {}

func (w *writeToControl)info(msg string){
	fmt.Println(msg)
}

type writeToFile struct {
	filePath string
}

func (w *writeToFile)info(msg string){
	fileObj, err := os.OpenFile(w.filePath, os.O_CREATE|os.O_APPEND, 0644)
	if err != nil{
		fmt.Printf("Write File Err: %v\n", err)
		return
	}
	fileObj.WriteString("info" + msg + "\n")
}

func write(msg string, arg ...writer)  {
	for _, obj := range arg{
		obj.info(msg)
	}
}

func main() {
	wc := &writeToControl{}
	wf := &writeToFile{filePath:"log.txt"}

	write("测试写入数据！", wc, wf)
}

**Comment by wilsonliu123** [comment]
sample code 
```
package main

import (
	"fmt"
	"io"
	"os"
)

type Logger interface {
	Output(w io.Writer)
}
type LogFile struct {
}

func (l *LogFile) Output(w io.Writer) {
	fmt.Fprint(w, "log message to file ")
}

type LogScreen struct {
}

func (l *LogScreen) Output(w io.Writer) {
	fmt.Fprint(w, "log message to screen")

}
func main() {
	var x Logger
	x = &LogFile{}
	f, err := os.Create("test.log")
	if err != nil {
		panic(err)

	}
	defer f.Close()
	x.Output(f)

	x = &LogScreen{}
	x.Output(os.Stdout)

}

```

**Comment by biuaxia** [comment]
我想请问下，这样子虽然可以确保接口被实现。

但是实际业务中有什么应用场景呢？

```golang
// 摘自gin框架routergroup.go
type IRouter interface{ ... }

type RouterGroup struct { ... }

var _ IRouter = &RouterGroup{}  // 确保RouterGroup实现了接口IRouter
```

声明出来的的 `_ IRouter` 这个也没有办法直接使用呢。

**Comment by biuaxia** [comment]
我大致的阅读了一下 `gin` 中 `routergroup.go` 同级的

- response_writer.go
- gin.go

这些文件，大致知道了实际用法；也就一个。

> 编写代码时加上这一句可以确保 `RouterGroup` 一定实现了接口 `IRouter`

其余好像没啥用来着，也并未在 `gin` 项目中找到其它用处。

但我还是想要知道 `var _ IRouter = &RouterGroup{}` 这个声明出来还有其它用处吗？

**Comment by DaiYifan** [comment]
您好，我看作用中，“你会发现main函数中明显有重复的代码，如果我们后续再加上猪、青蛙等动物的话，我们的代码还会一直重复下去。那我们能不能把它们当成“能叫的动物”来处理呢？” 我觉得这一块内容是想说复用性？这个听起来像是继承的作用，而不是接口的作用吧？即使用了接口，也要把一个个对象实例化 ，赋值给接口变量，并没有精简多少代码

**Comment by Q1mi** [comment]
> @biuaxia
> 我大致的阅读了一下 `gin` 中 `routergroup.go` 同级的
> 
> - response_writer.go
> - gin.go
> 
> 这些文件，大致知道了实际用法；也就一个。
> 
> > 编写代码时加上这一句可以确保 `RouterGroup` 一定实现了接口 `IRouter`
> 
> 其余好像没啥用来着，也并未在 `gin` 项目中找到其它用处。
> 
> 但我还是想要知道 `var _ IRouter = &RouterGroup{}` 这个声明出来还有其它用处吗？

确保接口被实现，把问题暴露在编译阶段。

**Comment by Q1mi** [comment]
> @DaiYifan
> 您好，我看作用中，“你会发现main函数中明显有重复的代码，如果我们后续再加上猪、青蛙等动物的话，我们的代码还会一直重复下去。那我们能不能把它们当成“能叫的动物”来处理呢？” 我觉得这一块内容是想说复用性？这个听起来像是继承的作用，而不是接口的作用吧？即使用了接口，也要把一个个对象实例化 ，赋值给接口变量，并没有精简多少代码

引出为什么要使用接口这里确实有点牵强，我回头修改下这里的内容。
感谢反馈。

**Comment by biuaxia** [comment]
> > @biuaxia
> > 我大致的阅读了一下 `gin` 中 `routergroup.go` 同级的
> > 
> > * response_writer.go
> > * gin.go
> > 
> > 这些文件，大致知道了实际用法；也就一个。
> > > 编写代码时加上这一句可以确保 `RouterGroup` 一定实现了接口 `IRouter`
> > 
> > 
> > 其余好像没啥用来着，也并未在 `gin` 项目中找到其它用处。
> > 但我还是想要知道 `var _ IRouter = &RouterGroup{}` 这个声明出来还有其它用处吗？
> 
> 确保接口被实现，把问题暴露在编译阶段。

好的，已经了解了，谢谢告知。

**Comment by DaiYifan** [comment]
> @Q1mi
> > @DaiYifan
> > 您好，我看作用中，“你会发现main函数中明显有重复的代码，如果我们后续再加上猪、青蛙等动物的话，我们的代码还会一直重复下去。那我们能不能把它们当成“能叫的动物”来处理呢？” 我觉得这一块内容是想说复用性？这个听起来像是继承的作用，而不是接口的作用吧？即使用了接口，也要把一个个对象实例化 ，赋值给接口变量，并没有精简多少代码
> 
> 引出为什么要使用接口这里确实有点牵强，我回头修改下这里的内容。
> 感谢反馈。

谢谢，通过您的博客学到了很多，点赞👍

**Comment by QuYingying0628** [comment]
1. qyyLog包
```go
package qyyLog

import (
	"fmt"
	"io"
	"os"
)

// 定义接口
type PrintLoger interface {
	PrintLog(logString string)
}

// 定义向终端打印日志的接口
type Terminal struct{}

// 定义往日志写文件的接口
type File struct {
	// 定义需要写入的文件名
	LogFileName string
}

// 终端实现这个打印日志接口
func (t Terminal) PrintLog(logString string) {
	fmt.Println(logString)
}

// 文件实现这个打印日志接口
func (f *File) PrintLog(logString string) {
	// 定义文件类型变量
	var file *os.File

	// 先判断这个文件是否存在
	if CheckFileIsExist(f.LogFileName) {
		// 文件存在时，打开文件
		file, _ = os.OpenFile(f.LogFileName, os.O_APPEND|os.O_WRONLY, 0666)
	} else {
		// 文件不存在是，创建文件
		file, _ = os.Create(f.LogFileName)
	}

	// 最后关闭文件（文件有打开就有关闭）
	defer file.Close()

	// 开始写入数据
	n, err := io.WriteString(file, logString+"\n")
	if err != nil {
		panic(err)
	}
	fmt.Printf("写入 %d 个字节\n", n)
}

// 判断文件是否存在的方法（看评论区大神的）
func CheckFileIsExist(filename string) bool {
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		return false
	}
	return true
}
```
2. main
```go
// 定义终端结构体
	terminal := qyyLog.Terminal{}
	// 定义文件结构体
	file := qyyLog.File{
		LogFileName: "log.txt",
	}

	// 定义接口
	var printer qyyLog.PrintLoger
	// 终端打印日志
	printer=terminal
	printer.PrintLog("永无bug！")
	// 文件打印日志
	printer=&file
	printer.PrintLog("永无bug！")
```

**Comment by bluesbonewong** [comment]
说一下，最后一小节是switch不是一般的switch语句，而是type switch
是个语法糖
博主可以注明一下，不然新手读完会困惑

**Comment by kikyoulg** [comment]
怎么套来套去的，看起来很麻烦

**Comment by lautaromartinez8964** [comment]
// 练习九：使用接口及依赖注入的方式实现一个既可以往终端写日志也可以往文件写日志的简易日志库

// 1.契约层(Abstraction): Logger 接口
//
//	LOgger定义了所有日志记录器都必须遵守的规范
type Logger interface {
	Log(message string)
	Error(message string)
}

// 2.实现层(Implementations):具体的日志记录器
// ---控制台日志实现--
type ConsoleLogger struct{} // ConsoleLogger将日志打印到标准输出(控制台)

func NewConsoleLogger() ConsoleLogger {
	// ConsoleLogger的构造函数
	return ConsoleLogger{}
}

func (c ConsoleLogger) Log(message string) {
	log.Printf("CONSOLE LOG:%s\n", message)
}

func (c ConsoleLogger) Error(message string) {
	log.Printf("CONSOLE ERROR:%s\n", message)
}

// ---文件日志实现---
type FileLogger struct {
	file *os.File // file变量存储的是一个指向os.File的指针 文件句柄
}

func NewFileLogger(filename string) (*FileLogger, error) {
	// FileLogger的构造函数，会打开或创建指定的日志文件，如果失败则返回错误
	// os.O_CREATE: 文件不存在则创建 os.O_WRONLY: 只写模式 os.O_APPEND: 追加内容到文件末尾
	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
	if err != nil {
		return nil, err
	}

	return &FileLogger{file: file}, err
	// 任何可能失败的初始化操作（如打开文件，连接数据库）都应当在构造函数中处理，并通过返回一个error来明确告知调用者操作是否成功

}

// 将日志写入文件的函数，使用指针接收者，因为它需要操作结构体内部的file字段
func (f *FileLogger) Log(message string) {
	// 使用fmt.Fprintf将格式化后的字符串写入到f.file中
	fmt.Fprintf(f.file, "FILE LOG:%s\n", message)
}

func (f *FileLogger) Error(message string) {
	fmt.Fprintf(f.file, "FILE Error:%s\n", message)
}

func (f *FileLogger) close() {
	f.file.Close() // Close是一个重要的办法，用于在程序结束时关闭文件句柄，防止资源泄漏
}

// 3.消费层 依赖于接口的服务
// UserService 负责用户相关的业务逻辑，它依赖于Logger接口 (业务服务，依赖抽象接口)
type UserService struct {
	logger Logger
}

func NewUserService(logger Logger) *UserService {
	// 业务服务的构造函数，通过依赖注入来接受一个Logger
	return &UserService{logger: logger}
}

func (us *UserService) CreateUser(username string) {
	//
	us.logger.Log(fmt.Sprintf("starting to create user'%s'...", username))
	// 这里是创建用户的复杂逻辑
	fmt.Printf("... (业务逻辑) User '%s' created in database.\n", username)
	us.logger.Log(fmt.Sprintf("User '%s' created successfully.", username))
}

func (us *UserService) DeleteUser(username string) {
	us.logger.Error(fmt.Sprintf("Starting to delete user '%s'...", username))
	// ... 假设这里是危险的删除用户逻辑 ...
	fmt.Printf("... (业务逻辑) User '%s' deleted from database.\n", username)
	us.logger.Error(fmt.Sprintf("User '%s' deleted successfully.", username))

}

**Comment by seabra2** [comment]
package main

import (
	"fmt"
	"os"
	"time"
)

// 使用接口的方式实现一个既可以往终端写日志也可以往文件写日志的简易日志库。

// LogWriter 日志输出契约
type LogWriter interface {
	Write(message string) error
}

type ConsoleWriter struct{}

func (c *ConsoleWriter) Write(message string) error {
	_, err := fmt.Fprintln(os.Stdout, message)
	return err
}

type FileWriter struct {
	file *os.File
}

func NewFileWriter(path string) (*FileWriter, error) {
	// os.O_APPEND: 追加写入, os.O_CREATE: 不存在则创建
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return nil, err
	}
	return &FileWriter{file: f}, nil
}

func (f *FileWriter) Write(message string) error {
	_, err := fmt.Fprintln(f.file, message)
	return err
}

// 组装日志库 (The Logger)
type Logger struct {
	Writer LogWriter // 核心：接口类型的插槽
}

func (l *Logger) Info(msg string) {
	now := time.Now().Format("2006-01-02 15:04:05")
	fullMsg := fmt.Sprintf("[%s] [INFO] %s", now, msg)

	// 调用接口方法
	if l.Writer != nil {
		l.Writer.Write(fullMsg)
	}
}

func main() {
	// 1. 初始化一个终端日志器
	console := &ConsoleWriter{}
	log := &Logger{Writer: console}
	log.Info("这条日志会出现在屏幕上")

	// 2. 随时切换为文件日志器
	file, err := NewFileWriter("app.log")
	if err != nil {
		fmt.Println("打开文件失败:", err)
		return
	}
	defer file.file.Close() // 记得关闭文件

	log.Writer = file // 像换电池一样切换输出源
	log.Info("这条日志会安静地躺在 app.log 文件里")
}


主要是插拔式的设计，大多数包都是按照这种想法设计，根据你传入的参数进行修改
